cpp/src/structure/remove_self_loops_impl.cuh

```diff
@@ -0,0 +1,180 @@
+/*
+ * Copyright (c) 2023, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#pragma once
+
+#include <raft/core/handle.hpp>
+#include <rmm/device_uvector.hpp>
+
+#include <thrust/distance.h>
+#include <thrust/iterator/zip_iterator.h>
+#include <thrust/remove.h>
+#include <thrust/tuple.h>
+
+#include <algorithm>
+#include <optional>
+
+namespace cugraph {
+
+namespace detail {
+
+template <typename vertex_t>
+void remove_self_loops(raft::handle_t const& handle,
+                       rmm::device_uvector<vertex_t>& edgelist_srcs /* [INOUT] */,
+                       rmm::device_uvector<vertex_t>& edgelist_dsts /* [INOUT] */)
+{
+  auto edge_first =
+    thrust::make_zip_iterator(thrust::make_tuple(edgelist_srcs.begin(), edgelist_dsts.begin()));
+  edgelist_srcs.resize(
+    thrust::distance(
+      edge_first,
+      thrust::remove_if(handle.get_thrust_policy(),
+                        edge_first,
+                        edge_first + edgelist_srcs.size(),
+                        [] __device__(auto e) { return thrust::get<0>(e) == thrust::get<1>(e); })),
```

Comment:
I am 70% sure thrust::remove_if requires 2x the input array size memory. If this becomes a memory bottleneck (e.g. the input edge list size is large compared to the device memory size, see https://github.com/rapidsai/cugraph/blob/branch-23.12/cpp/src/structure/create_graph_from_edgelist_impl.cuh#L523C23-L523C23), we may create a bitmap storing whether to remove an edge or not. And based on this bitmap, we can call remove_if on srcs, dsts, weights, edge_ids, and edge_types, separately. Then, we can keep the peak memory requirement under our goal.

Reply:
That would also eliminate the need for the ugliness you mention below.  I'll refactor.